src/base64.cpp

```diff
@@ -152,60 +163,128 @@ std::string encode( const char* begin, const char* end, int newlines)
 	// add the partial group
 	if(extra)
 	{
-		uint32_t copy = 0;
-		uint8_t* bits = reinterpret_cast<uint8_t*>(&copy);
+		std::uint8_t partial_group[4] = { 0 };
 		for(unsigned i = 0; i < tail; ++i)
-			bits[i] = *(read++);
+			partial_group[i] = *(read++);
+		const uint8_t* const_partial_group = partial_group;
+		encode(const_partial_group, write);
 
-		const uint8_t* cp = bits;
-		encode(cp, write);
+		buffer[buffer.length() - 1] = '=';
+		if (tail == 1) 
+		{
+			buffer[buffer.length() - 2] = '=';
+		}
 	}
-
+	
 	assert( (char*)read == end );
-	assert( write == buffer.end() - tail );
-
+	assert( write == buffer.end());
+	
 	return buffer;
 }
 
-void decode( std::uint8_t*& byte_array, std::string::const_iterator& reader )
-{
-	uint32_t* bits = reinterpret_cast<uint32_t*>(byte_array);
-	*bits &= 0;
-	for(int i = 0; i < 4; ++i)
+std::uint8_t decode_byte_one(const char* reader) {
+	std::uint8_t res;
+	uint8_t const value0 = decode(reader[0]);
+	uint8_t const value1 = decode(reader[1]);
+	if( ((value0 & (~bitmask)) != 0 ) ||
+	    ((value1 & (~bitmask)) != 0 ) )
 	{
-		uint8_t value = decode(*(reader++));
-		assert( (value & (~bitmask)) == 0 ); // check that it is valid
-		*bits |= value << (6*i);
+		throw std::runtime_error("invalid base64 character");
 	}
-	std::advance(byte_array, 3);
+	res = (value0 << 2);
+	res |= value1 >> 4;
+
+	return res;
+}
+
+std::uint8_t decode_byte_two(const char* reader) {
+	std::uint8_t res;
+	uint8_t const value1 = decode(reader[1]);
+	uint8_t const value2 = decode(reader[2]);
+	if( ((value1 & (~bitmask)) != 0 ) ||
+	    ((value2 & (~bitmask)) != 0 ) )
+	{
+		throw std::runtime_error("invalid base64 character");
+	}
+	res = value1 << 4;
+	res |= value2 >> 2;
+
+	return res;
 }
 
+std::uint8_t decode_byte_three(const char* reader) {
+	std::uint8_t res;
+	uint8_t const value2 = decode(reader[2]);
+	uint8_t const value3 = decode(reader[3]);
+	if( ((value2 & (~bitmask)) != 0 ) ||
+	    ((value3 & (~bitmask)) != 0 ) )
+	{
+		throw std::runtime_error("invalid base64 character");
+	}
+	res = value2 << 6;
+	res |= value3;
+
+	return res;
+}
```

Comment:
is this function still used anywhere? If not, should we delete it?

Reply:
Removed.